text/0000-int-guidelines.md

```diff
@@ -0,0 +1,164 @@
+* 32-bit integers are considerably faster than 64-bit integers in some
+  situations, so using 64-bit integers for tiny numbers, especially in
+  hot code, can result in unnecessarily slower programs.
+* Reflexive use of 32-bit integers too-often results in overflows when the
+  numbers aren't expected to be "laughably smaller" than the maximum
+  32-bit number.
+* Occasional use of `usize` when building a brand new data structure may
+  be appropriate, but use of `usize` in general can introduce portability
+  hazards when the use-case is not proportional to the amount of
+  addressable memory.
+* Using of unsigned integers is traditionally thought to be error-prone, and
+  style guides often suggest avoiding them. That said, Rust's unsigned integers
+  have built-in underflow assertions, which changes the analysis.
+
+# Detailed design
+
+When deciding what integer type to use for a given situation, you can make the decision
+to a first approximation through this heuristic:
+
+* If you expect all uses of this API to use numbers "laughably smaller" then 32-bits,
+  use 32-bit numbers. Otherwise, prefer 64-bit numbers.
+* If you are sure that the number will never be less than 0, use unsigned integers.
+  Otherwise, use signed numbers.
+* If you are building a new data structure, and your number refers to the size of memory,
+  you may want to use `usize`, which is described below.
+
+> Note: Rust does not yet have BigInts in `std`, which might, in theory, be a better
+> go-to big integer type. These guidelines will probably be revised in the future
+> once that changes.
+
+## Signed vs. Unsigned Integers
+
+Traditionally, style guides for low-level languages have [warned against the use
+of unsigned integers](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Integer_Types) because
+they can introduce bugs when comparing unsigned value with 0
+(`for (unsigned int i = foo.len() - 1; i >= 0; --i) ...`). The Google Style Guide
+argues that instead of unsigned integers, programmers should use signed
+integers and assertions that the value does not go below 0.
+
+In Rust, unsigned integers have underflow checking assertions built-into
+the type (assuming that RFC XXX is accepted), so using a `u32` is equivalent
+to the advice in Google style guide (with a larger maximum value).
+
+## The "size" Types
+
+The `isize` and `usize` types represent numbers that are proportional to
+the size of available memory. Most normal uses of `isize` and `usize`
+should arise through the use of existing data structures that expose
+values with those types.
+
+For example, if a new structure wanted to store the length of a `Vec` in
+one of its fields, it should store it as a `usize`, because that's the return
+value of `vec.len()`.
+
+When building entirely new low-level data structures, you may want to use
+`usize` to represent a value (e.g. node id) that scales linearly with the amount
+of addressable memory.
+
```

Comment:
A common use case for usize is for representing pointers as numbers. E.g., if you want to do pointer arithmetic, then pointers are usize and offsets are isize.

Also, indexes into arrays.

These two should cover the more abstract concept of 'scaling with addressable memory'